### PR TITLE
rows as panels in dashboard; gridpos for row, graph, text panels

### PR DESCRIPTION
**Pull request by victoriaalee**
## What does this do?
Rows are now considered as panels in the dashboard. The attributes of the `Row`, `Graph`, and `Text` panels are also updated; the main change is adding `GridPos`.

## Why is it a good idea?
Rows as panels and `GridPos` were introduced several dashboard schema versions ago.

## Context
I created a `Panel` class, which the `Row`, `Graph`, and `Text` panels subclass.

Grafana source code referenced:

- https://github.com/grafana/grafana/blob/v6.0.0/public/app/features/dashboard/state/DashboardModel.ts
- https://github.com/grafana/grafana/blob/v6.0.0/public/app/features/dashboard/state/PanelModel.ts
- https://github.com/grafana/grafana/blob/v6.0.0/public/app/plugins/panel/graph/module.ts
- https://github.com/grafana/grafana/blob/v6.0.0/public/app/plugins/panel/text/module.ts

## Questions
N/A

### CODE REVIEW

**Comment by bboreham** [comment]
Thanks for the PR.  

There seem to be a few names like `datasource` and `timepicker` where the case has changed.  Is it essential to break backwards-compatibility in this way?

**Comment by victoriaalee** [comment]
Just pushed commits to revert those changes. (I altered them because they better fit the code style at my company.)

**Comment by bboreham** [comment]
Thanks for that, it's much easier to review now.
As with #141, I have a reaction to needing to edit every single one of my existing dashboards to replace `rows` - is there some way to do it without that?

**Comment by victoriaalee** [comment]
Hi @bboreham,

I did a couple rough tests by manually editing JSON in the Grafana UI - these shouldn't be taken as source of truth by any means but might give us some insight.

With Dashboards still being made up of `rows` (not `panels`):
- **specifying `schemaVersion` 12** - GridPos info in the JSON panel object is disregarded in rendering of Grafana
- **specifying `schemaVersion` 20** - nothing shows up; this should be an issue with versions 14+

So it seems that keeping Dashboards with `rows` is incompatible with `GridPos`

My personal opinion is that it is appropriate to release a new grafanalib version that supports the latest schema version (with GridPos) instead of trying to maintain backwards compatibility. I think that if people desire to use GridPos, they should be upgrading to use the new schema version and the new way of defining dashboards (with rows counting as panels). Just my two cents :)

Let me know your thoughts on how to proceed with this - I'd really like to see some form of GridPos be added to grafanalib in the near future, whatever way it may look like!

File: grafanalib/core.py [hunk 1]
```diff
@@ -460,46 +478,60 @@ def _balance_panels(panels):
...
-            'collapse': self.collapse,
-            'editable': self.editable,
-            'height': self.height,
-            'panels': self.panels,
-            'showTitle': showTitle,
-            'title': title,
+            'collapsed': self.collapsed,
+            'gridPos': self.gridPos,
+            'id': self.id,
+            'panels': self.panels if self.collapsed else [],
```
Comment (robobeaver6) [inline grafanalib/core.py]: I am not sure why this conditional is here, but if Collapsed is not specified or it is set to False, then the JSON dumps has empty panels. I am not sure why you would want this behaviour.

**Comment by robobeaver6** [comment]
I really want this gridPos functionality.  Thanks for the PR.  I am testing this now and I noticed  [in line531](https://github.com/victoriaalee/grafanalib/blob/27f4f67147be8a6a52d950a9043f2981fae6b787/grafanalib/core.py#L531) there is a condition to only return self.panels if self.collapsed is true.  This results in the JOSN dump not adding any panels for non collapsed Rows.  I can not think of a reason you would want to do this.  Althoug on removing it, I notice that the Dashboard does not populate the Graphs and Targets even though they are in the JSON data.

Was there a reason the sharedCrosshair variable removed?  I quite like the functionality, and if seams to work fine if I add it back in.

Also as a side note,  by having the Row, Graph and Text classes inherit from Panel, it breaks Pycharm's hints.  I get a lot of "uexpected argument" warnings because it doesn't seem to realise it needs to parse the parent object.  Not sure if this is Pycharm being broken, or I need to set something different.

File: grafanalib/core.py (hunk 1 above)
Comment (victoriaalee) [reply grafanalib/core.py]: I observed this behavior from playing with the Grafana UI and checking the generated JSON. It isn't intuitive to me either, but it seems like this is how Grafana handles collapsed rows.

**Comment by victoriaalee** [comment]
Replied to your original comment about collapsed rows.

I removed `sharedCrosshair` because it seems that that is now handled with the `graphTooltip` field. Specifically, if you use the UI to select `Shared crosshair` for the value of `Graph Tooltip`, you will see that the value of the JSON field `graphTooltip` changes to `1`. Is this shared crosshair graph tooltip the functionality that you were hoping to maintain?

Hm, I'm not seeing "unexpected argument" warnings on my end. I use IntelliJ with the Python Plugin and configured a Python 3.6 virtualenv with installed dependencies (`pip install -e .`).

**Comment by matthewmrichter** [comment]
My org is capturing a bunch of dashboards right now and having gridPos would make life a lot easier. Anything I could do to help get this to a mergeable state? Happy to be a guinea pig as well.

**Comment by robobeaver6** [comment]
This thread seems to have gone a bit cold.  Is it likely to be merged any time soon?  Seeing as is a function of the current schema, it seems sensible to add this functionality.  Is there anything that needs doing or changing to get it merged?  GridPos is fairly essential to laying out dashboards with anything other than basic complexity.  Happy to help in any way I can to get this functionality

**Comment by victoriaalee** [comment]
Also would still love to see this merged in!

**Comment by matthewmrichter** [comment]
Mostly we haven't put much testing or thought into this yet just because we're working on low-hanging fruit as we catch up. We would like to see this functionality but also this looks like it could break compatibility. Maybe you can help us understand the impact of this change on existing dashboard code?

**Comment by dholbach** [comment]
Just a quick PSA from me: 
1. we're hanging out in [#grafanalib on Slack](https://weave-community.slack.com/archives/grafanalib)
1. we're planning a contributors meeting soon: here's the [link to the doodle](https://doodle.com/poll/hgsp9wmuff7m9bi7) to find a good time - be sure to pick your timezone before submitting ([preliminary agenda](https://docs.google.com/document/d/1JxrSszyPHYhNbJDWYZehRKv6AO4U-zIBhuNmYQVOIHo/edit))

We'd love to see you there.

**Comment by Agitolyev** [comment]
Hi, first of all, I'd like to say that this project helps a lot to manage grafana dashboards. Many thanks to all of you. The only thing that stops us from completely moving to manage our dashboards as a code is that we need to manipulate panels' positions. And that PR contains precisely that functionality, but it seems to cool down and now it is labeled as WIP, maybe you have some rough estimates for when it will be available in main, and/or there is some assistance needed ?

**Comment by JamesGibo** [comment]
Hi @Agitolyev 
If you have sometime to help that would be greatly appreciated as this is a feature that many people are waiting on.

The main issue with this PR is that the code has changed quite a bit underneath it. 
Support for it needs switching to the new generic panel class
https://github.com/weaveworks/grafanalib/blob/f071d2c5415894e3ebfe44d94bfe3a30b5f6eb5e/grafanalib/core.py#L1089

Another attempt at adding support for grid position and the grafana schema version 18 is in a branch
https://github.com/weaveworks/grafanalib/compare/schemaVersion18

Ideally, it needs to be added in a way that maintains backwards compatibility with existing grafana schema versions, which is the bit I am a bit unsure about and will require some testing. 

In the future it would be good to be able to specify the schema  version you require and generate the correct json
https://github.com/grafana/grafana/tree/master/dashboard-schemas
https://grafana.com/docs/grafana/latest/dashboards/json-model/

**Comment by kevin-s-wang** [comment]
I'm also trying to add the `gridPos` functionality for our projects, however, I've found this PR and all the functionalities I want are in it. Hope this PR could soon be merged.

**Comment by charlesoconor** [comment]
This is great. Going to set up my attempt to use this lib off of the branch so hope it's merged since it is necessary functionality.

**Comment by charlesoconor** [comment]
Any movement on this or is this repo dead?

**Comment by JamesGibo** [comment]
I created a new PR based off this one to add support for girdPos to Grafanalib as the master branch has diverged too much from this PR to make a rebase practical. 
#319 
I will be working on this PR over the next week to get it ready to be merged, if anyone is able to test it and check it has all the features they require, this would be much appreciated.

**Comment by charlesoconor** [comment]
> I will be working on this PR over the next week to get it ready to be merged, if anyone is able to test it and check it has all the features they require, this would be much appreciated.

I pulled the branch and it seemed to work for me. Given my grafanlib usage is very new.

**Comment by JamesGibo** [comment]
Closing this PR as most of the features have been merged in #319

**Pull request closed without merging**